README.md

```diff
@@ -18,38 +18,271 @@ sources).
+
+* [kas-container 3.0.2](https://raw.githubusercontent.com/siemens/kas/3.0.2/kas-container)
+  script downloaded and available in [PATH](https://en.wikipedia.org/wiki/PATH_(variable))
+
+    ```bash
+    wget -O ~/bin/kas-container https://raw.githubusercontent.com/siemens/kas/3.0.2/kas-container
+    chmod +x ~/bin/kas-container
+    ```
+
+* `meta-trenchboot` repository cloned
+
+    ```bash
+    mkdir yocto
+    cd yocto
+    git clone https://github.com/3mdeb/meta-trenchboot.git
+    ```
+
+* [bmaptool](https://docs.yoctoproject.org/dev-manual/bmaptool.html) installed
+
+    ```bash
+    sudo dnf install bmap-tools
+    ```
+
+    > You can also use `bmap-tools`
+    > [from github](https://github.com/yoctoproject/bmaptool) if it is not
+    > available in your distro.
+
+## Build
+
+* From `yocto` directory run:
+
+    ```shell
+    kas-container build meta-trenchboot/kas-generic-tb.yml
+    ```
+
+* Image build takes time, so be patient and after build's finish you should see
+something similar to (the exact tasks numbers may differ):
+
+    ```shell
+    Initialising tasks: 100% |###########################################################################################| Time: 0:00:01
+    Sstate summary: Wanted 360 Found 0 Missed 8 Current 1810 (0% match, 99% complete)
+    NOTE: Executing Tasks
+    NOTE: Tasks Summary: Attempted 4774 tasks of which 4749 didn't need to be rerun and all succeeded.
+    ```
+
+Thanks to publishing the build cache on cache.dasharo.com the time needed to
+finish the process should be significantly decreased.
+Using the cache is enabled in kas/cache.yml file and can be disabled by removing
+reference to this file in `kas/common.yml`:
+
+```yaml
+includes:
+    - cache.yml
+```
+
+This cache can decrease time needed to build image from scratch from hours to
+minutes depending on build machine and network connection.
+
+```shell
+Sstate summary: Wanted 2170 Local 0 Mirrors 2151 Missed 19 Current 0 (99% match, 0% complete)
+NOTE: Executing Tasks
+NOTE: Tasks Summary: Attempted 4774 tasks of which 4445 didn't need to be rerun and all succeeded.
+```
+
+## Flash
+
+To flash resulting image:
+
+* Find out your device name:
+
+    ```shell
+    $ lsblk
+    NAME                                     MAJ:MIN RM   SIZE RO TYPE  MOUNTPOINTS
+    sdx                                      179:0    0  14.8G  0 disk
+    ├─sdx1                                   179:1    0     4M  0 part
+    └─sdx2                                   179:2    0     4M  0 part
+    ```
+
+    In this case the device name is `/dev/sdx` **but be aware, in next steps
+    replace `/dev/sdx` with right device name on your platform or else you can
+    damage your system!.**
+
+* From where you ran image build type:
+
+    ```shell
+    cd build/tmp/deploy/images/genericx86-64/
+    sudo umount /dev/sdx*
+    sudo bmaptool copy tb-minimal-image-genericx86-64.rootfs.wic.gz /dev/sdx
```

Comment:
Is it possible to copy just one component (e.g. SKL or GRUB) after rebuilding, without flashing whole image? This is very time-consuming, and will be needed often during development.

Reply:
Yes (somewhat).
Easiest way would be to create image with bmaptool, create loop device `udisksctl loop-setup -f trenchboot.img`, mount partition with changes and copy them to flash drive or via ssh (rsync).
You could also find built files in `build/tmp/work/<arch>/<recipe>/<version>/{deploy-<recipe>|image}`
where `<arch>` is `core2-64-tb-linux` for `skl`, `grub`, `grub-efi`, `intel-sinit-acm` and `genericx86_64-tb-linux` for `linux-tb` but you need to know which files you want and where they should go.
Depending on which files you want it's possible all of them would be in `build/tmp/deploy/images/genericx86-64`.
There is also [devtool deploy-target](https://docs.yoctoproject.org/ref-manual/devtool-reference.html#deploying-your-software-on-the-target-machine) which could be used to copy files created in `do_install` task of recipe by ssh. Tested on my local machine:

```shell
devtool deploy-target -s grub-efi user@ip:/home/user/devtool/grub-efi
```

And it copied:

```shell
~ tree -L 4 ~/devtool/grub-efi
/home/user/devtool/grub-efi
├── boot
│   └── EFI
│       └── BOOT
│           └── bootx64.efi
└── usr
    └── lib
        └── grub
            └── x86_64-efi
```

Could be used if target machine can boot into linux, but as you can see not every file is copied (no `grub.cfg`).

Reply:
> Easiest way would be to create image with bmaptool, create loop device udisksctl loop-setup -f trenchboot.img, mount partition with changes and copy them to flash drive or via ssh (rsync).

It does not sound easy at all. If you want to extract stuff from image, you should be able to extract them from yocto working directories as well.

> There is also [devtool deploy-target](https://docs.yoctoproject.org/ref-manual/devtool-reference.html#deploying-your-software-on-the-target-machine) which could be used to copy files created in do_install task of recipe by ssh. Tested on my local machine:


Devtool is a proper proposal here. You should not need to provide any path in arguments, simply user and IP address.

But would it work also for SKL and kernel properly?
grub.cfg is a simple file and can be added to the script, if we use scp anyway

The grub legacy might be a bit more problematic, as I suppose devtool would not reinstall grub on disk

Reply:
@macpijan I meant easy in way that there is already proper directory structure that you can copy (and could be made faster with scripts).

> You should not need to provide any path in arguments

This was just a test to see which files it would copy to my own local machine, didn't want it to overwrite important files.

> But would it work also for SKL and kernel properly?

Probably not, at least not completely. I think `skl.bin` is put on boot partition during `do_wic` task.
Rootfs files are probably ok but there are problems if something should be on `boot` partition, either it isn't or there are some different files. 

```shell
$ tree -d -L 6
.
├── linux-tb
│   ├── boot
│   └── usr
│       └── lib
│           └── modules
│               └── 6.6.1-yocto-standard
│                   └── kernel
└── skl
    └── usr
        └── bin
            └── skl
```

```shell
& tree linux-tb/boot -h
[  298]  linux-tb/boot
├── [  11M]  bzImage-6.6.1-yocto-standard
├── [ 152K]  config-6.6.1-yocto-standard
├── [ 751K]  Module.symvers-6.6.1-yocto-standard
├── [ 6.7M]  System.map-6.6.1-yocto-standard
└── [  41M]  vmlinux-6.6.1-yocto-standard
```

From what I can see it's only kernel without initramfs.